src/main/kotlin/org/rust/lang/refactoring/RsImportOptimizer.kt

```diff
@@ -72,18 +71,63 @@ class RsImportOptimizer : ImportOptimizer {
             val first = mod.childrenOfType<RsElement>()
                 .firstOrNull { it.textOffset >= offset && it !is RsExternCrateItem && it !is RsAttr } ?: return
             val psiFactory = RsPsiFactory(mod.project)
-            val sortedUses = uses
-                .sortedBy { it.useSpeck?.pathText }
-                .mapNotNull { it.copy() as? RsUseItem }
-            sortedUses
-                .mapNotNull { it.useSpeck }
-                .forEach { optimizeUseSpeck(psiFactory, it) }
-            for (importPath in sortedUses) {
-                mod.addBefore(importPath, first)
+            val sortedUsesGroups = uses
+                .map { UseItemWrapper(it) }
+                .groupBy({ it.packageGroupLevel }, { it })
+                .map { (groupLevel, useItemWrapper) ->
+                    groupLevel to useItemWrapper
+                        .sortedBy { it.useSpeckText }
+                        .mapNotNull { it.useItem.copy() as? RsUseItem }
+                }
+                .filter { it.second.isNotEmpty() }
+                .sortedBy { it.first }
+            sortedUsesGroups.forEach { sortedUses ->
+                sortedUses.second
+                    .mapNotNull { it.useSpeck }
+                    .forEach { optimizeUseSpeck(psiFactory, it) }
+            }
+
+            for ((_, sortedUses) in sortedUsesGroups) {
+                val lastUseItem = sortedUses.map { importPath ->
+                    val addedUseItem = mod.addBefore(importPath, first)
+                    mod.addAfter(psiFactory.createNewline(), addedUseItem)
+                    addedUseItem
+                }.last()
+                mod.addAfter(psiFactory.createNewline(), lastUseItem)
+            }
+            uses.forEach {
+                (it.nextSibling as? PsiWhiteSpace)?.delete()
```

Comment:
Do we really need it? If yes, can we add the corresponding case in test?

Reply:
Yes, we need it
I have added corresponding [test](https://github.com/isamborskiy/intellij-rust/blob/af3a3a6607d5ed36994fba8bc82174a51c200677/src/test/kotlin/org/rust/lang/refactoring/RsImportOptimizerTest.kt#L364)